jdaviz/configs/specviz/plugins/viewers.py

```diff
@@ -362,10 +363,27 @@ def add_data(self, data, color=None, alpha=None, **layer_state):
         result : bool
             `True` if successful, `False` otherwise.
         """
+        y_units = data.get_component("flux").units
+
         # If this is the first loaded data, set things up for unit conversion.
         if len(self.layers) == 0:
             reset_plot_axes = True
         else:
+            # Check if the new data flux unit is actually compatible since flux not linked
+            data_flux_unit = u.Unit(y_units)
+            viewer_flux_unit = u.Unit(self.state.y_display_unit)
+            wav = 1 * u.Unit(self.state.x_display_unit)
+            try:
+                (1 * data_flux_unit).to(viewer_flux_unit, u.spectral_density(wav))  # Error if incompatible  # noqa: E501
```

Comment:
Technically you could also make use of whatever the default unit converter is in glue to do the 'test' unit conversion - just in case you ever modify your custom unit conversion class and it then becomes inconsistent with what is here. 

```python
from glue.core.units import UnitConverter
uc = UnitConverter()
uc.to_unit(data, "flux", 1, u.Unit(self.state.y_display_unit))
```

(untested)

Reply:
That didn't work out of the box. This does but I am not sure it is better or worse:

```python
uc.to_unit(data, data.find_component_id("flux"), [1, 1], u.Unit(sv.state.y_display_unit))
```

Reply:
> ```python
> uc.to_unit(data, "flux", 1, u.Unit(self.state.y_display_unit))
> ```
Jdaviz `UnitConverterWithSpectral needs an actual `cid` and a sequence for `values`, with 
```python
uc.to_unit(data, data.id["flux"], [1], viewer_flux_unit)
```
this would work. As a side note, I thought something like
```python
if viewer_flux_unit in [u.Unit(equ) for equ in uc.equivalent_units(data, data.id['flux'])]:
```
should also work (without having to try to run the actual unit conversion), but for the example data in the notebook I found this:
```pycon
data = viz.app.data_collection['NIRSpec']
print(uc.to_unit(data, data.id['flux'], [1], u.Jy / u.steradian))
print(u.Jy / u.steradian in [u.Unit(eqv) for eqv in uc.equivalent_units(data, data.id['flux'])])
print(data.get_component('flux').units)
print(uc.equivalent_units(data, data.id['flux']))

[1000000.]
False
MJy / sr
{'erg / (Angstrom s cm2)', 'erg / (s cm2 Angstrom)', 'W / (m2 Hz)', 'eV / (s m2 Hz)', 'erg / (s cm2 Hz)', 'ph / (s cm2 Angstrom)', 'ph / (Angstrom s cm2)', 'mJy', 'Jy', 'erg / (Hz s cm2)', 'ph / (Hz s cm2)', 'erg / (s cm2)', 'eV / (Hz s m2)', 'uJy', 'ph / (s cm2 Hz)', 'W / (Hz m2)'}
```
Is this something broken in `UnitConverterWithSpectral.equivalent_units`?

Reply:
The returned list of unit string is not exhaustive. I think in the end we still need to actually try to convert first because `astropy.units` does a bunch of stuff that just a list of string cannot tell us for sure.

Reply:
As for taking a list, that is also fine because internally, Specviz will never try to plot any spectrum with just a single flux value.

Reply:
> The returned list of unit string is not exhaustive. I think in the end we still need to actually try to convert first because `astropy.units` does a bunch of stuff that just a list of string cannot tell us for sure.

It is not just not exhaustive, but the present hardcoded list as it comes out of
https://github.com/spacetelescope/jdaviz/blob/fd8e3a92ce6bd20cbd5f2fe5e11761b58faab0c9/jdaviz/app.py#L79-L88
does not even contain its own unit if the 'flux' component in `data` is e.g. in `u.Jy / u.steradian`; so if that unit converter _should_ work with intensity spectra data, anything per sterad or deg**2 etc., this should probably be addressed.

But on comparing with the default glue converter, or what Astropy's `unit.find_equivalent_units()` actually returns, I realise it is indeed not sufficient to search that list. What one would actually need is the equivalent (pun-meter++) of Astropy's `unit.is_equivalent(viewer_flux_unit, equivalencies=eqv)`, but since that currently does not exist in the glue units framework. trying to do the actual conversion is probably the best option.

Reply:
I am not sure if we want to auto convert any `/sr` unit at all because it depends on pixel scale. The next loaded spectrum might not have the pixel scale info available or encoded in a way we know how to extract. It is a slippery slope.

I am also not sure how exhaustive `list_of_units` is supposed to be but it is quite unreasonable if `glue` expect devs to provide an exhaustive list to begin with.

Reply:
Not advocating for auto-converting anything, at least that would indeed have to be very carefully thought through.
But when working with Jy/sr data as in the example notebook, and the user provides a template in matching units, this should of course (and does) work. Yet `UnitConverterWithSpectral.equivalent_units()` will return a very misleading list for those data (whether it is actually used anywhere is another question).

> I am also not sure how exhaustive `list_of_units` is supposed to be but it is quite unreasonable if `glue` expect devs to provide an exhaustive list to begin with.

Rather to the contrary, from the example of `SimpleAstropyUnitConverter` `glue` only seems to expect the very basic list that `astropy.units` produces (which e.g. for MJy is only `['Jy', 'AB']`, or `['ST', 'Jy', 'bol', 'AB']` when including spectral_density equivalencies). Bringing back the question what its intended use is.

Reply:
> Bringing back the question what its intended use is.

Only @astrofrog knows.